File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
def backtrack(seq, visited):
    if len(seq) == M:
        print(' '.join(map(str,seq)))
        return
    
    for i in range(1, N+1):
        if not visited[i]:
            visited[i] = True
            seq.append(i)

            backtrack(seq, visited)

            seq.pop()
            visited[i] = False

N, M = map(int, input().split(' '))

visited = [False] * (N+1)

backtrack([], visited)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include <iostream>
#include <vector>

using namespace std;

int N, M;
vector<int> seq;
vector<bool> visited;

void backtrack() {
    if (seq.size() == M) {
        for (int num : seq) {
            cout << num << " ";
        }
        cout << "\n";
        return;
    }

    for (int i = 1; i <= N; ++i) {
        if (!visited[i]) {
            visited[i] = true;
            seq.push_back(i);

            backtrack();

            seq.pop_back();
            visited[i] = false;
        }
    }
}

int main() {
    cin >> N >> M;

    visited.resize(N + 1, false);

    backtrack();

    return 0;
}
